Привет! Как дела?) ☺️
<br> Меня зовут Дуолан 👋 Буду проверять твой проект. Давай вместе доведем его до идеала 😉
<br> Дальнейшее общение будет происходить на «ты», если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий («исправил», «не понятно как исправить ошибку», ...)
<br> Пожалуйста, не удаляй мои комментарии, они будут необходимы для повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>✔️ Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-warning">
<b>⚠️ Совет:</b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>❌ Замечание:</b> Если требуются исправления
</div>

Работа не может быть принята с красными комментариями.

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-info">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

<div class="alert alert-block alert-warning">
<b>⚠️ Совет:</b> 

Важно чтобы в проектах было подробное описание: когда потенциальные работодатели будут смотреть твой проект в портфолио, им нужно будет понимать задачу, чтобы можно было оценить её решение. Вот что можно добавить:
    
* Описание целей и задачи проекта
* Описание данных
* План работы

И рекомендую писать введение от первого лица множественного числа. Так работа выглядит более самостоятельной)
</div>

## Подготовка данных
<br>
<div style="margin: 10px 50px 10px 0px; padding: 20px; border: teal solid 1px ">
 Данные находятся в файле <code>/datasets/Churn.csv</code>
    </div>

<br>
<div style="margin: 10px 50px 10px 0px; padding: 20px; border: teal solid 1px ">
Признаки<br>
RowNumber — индекс строки в данных<br>
CustomerId — уникальный идентификатор клиента<br>
Surname — фамилия<br>
CreditScore — кредитный рейтинг<br>
Geography — страна проживания<br>
Gender — пол<br>
Age — возраст<br>
Tenure — сколько лет человек является клиентом банка<br>
Balance — баланс на счёте<br>
NumOfProducts — количество продуктов банка, используемых клиентом<br>
HasCrCard — наличие кредитной карты<br>
IsActiveMember — активность клиента<br>
EstimatedSalary — предполагаемая зарплата<br>
Целевой признак<br>
Exited — факт ухода клиента<br>
    </div>


In [1]:
import pandas as pd

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score
from sklearn.utils import shuffle

In [2]:
df = pd.read_csv('Churn.csv')

In [3]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [4]:
df.tail()

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
9995     5.0       0.00              2          1               0   
9996    10.0   57369.61              1          1               1   
9997     7.0       0.00              1          0               1   
9998     3.0   75075.31              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


<div style="margin: 10px 50px 10px 0px; padding: 20px; border: blue solid 1px ">
 В столбце Tenure есть пропуски. Их около 10%. Заполним их медианными значениями.
    </div>

<div class="alert alert-block alert-success">
<b>✔️ Успех:</b>

Ты хорошо владеешь методами первичного анализа данных 👍
</div>

In [6]:
df['Tenure'] = df['Tenure'].fillna(df['Tenure'].median())

# убедимся что пропусков не осталось
df['Tenure'] = df['Tenure'].astype(int)
df['Tenure'].value_counts()

5     1836
1      952
2      950
8      933
3      928
7      925
4      885
9      882
6      881
10     446
0      382
Name: Tenure, dtype: int64

<div class="alert alert-block alert-warning">
<b>⚠️ Совет:</b>

Посмотри еще на такие варианты работы с пропусками:

1. Распределение по годам использования банка практически ровное, если сейчас заменить 909 пропусков медианой, то мы увеличим в 2 раза значение для 5 лет. Это кажется некорректным. Можно сделать заполнение пропусков случайными числами, тогда мы снизим влияние на наше распределение.

2. Так как значений в этом признаке ограниченное количество, можно сделать его категориальным признаком. Пропуски можно считать как за отдельную категорию (заполнить значением -1). Затем заменить тип данных Tenure на object и применить технику OHE.

3. Заполнить "0", считая, что это новые клиенты.
</div>

<div style="margin: 10px 50px 10px 0px; padding: 20px; border: blue solid 1px ">
 Удалим столбцы которые не будут учавствовать в обучении модели.
    </div>

In [7]:
df = df.drop(columns=['RowNumber', 'CustomerId', 'Surname'])

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CreditScore      10000 non-null  int64  
 1   Geography        10000 non-null  object 
 2   Gender           10000 non-null  object 
 3   Age              10000 non-null  int64  
 4   Tenure           10000 non-null  int32  
 5   Balance          10000 non-null  float64
 6   NumOfProducts    10000 non-null  int64  
 7   HasCrCard        10000 non-null  int64  
 8   IsActiveMember   10000 non-null  int64  
 9   EstimatedSalary  10000 non-null  float64
 10  Exited           10000 non-null  int64  
dtypes: float64(2), int32(1), int64(6), object(2)
memory usage: 820.4+ KB


<div class="alert alert-block alert-success">
<b>✔️ Успех:</b>

Удалить неинформативные признаки - хорошее решение 👍
</div>

<div style="margin: 10px 50px 10px 0px; padding: 20px; border: blue solid 1px ">
 Закодируем признаки, чтобы категориальные данные стали численными
    </div>

In [8]:
df_ohe = pd.get_dummies(df)
df_ohe.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42       2       0.00              1          1   
1          608   41       1   83807.86              1          0   
2          502   42       8  159660.80              3          1   
3          699   39       1       0.00              2          0   
4          850   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  Geography_France  \
0               1        101348.88       1                 1   
1               1        112542.58       0                 0   
2               0        113931.57       1                 1   
3               0         93826.63       0                 1   
4               1         79084.10       0                 0   

   Geography_Germany  Geography_Spain  Gender_Female  Gender_Male  
0                  0                0              1            0  
1                  0                1              1            0  
2                  0                0              1            0  
3                  0                0              1            0  
4                  0                1              1            0

<div class="alert alert-block alert-danger">
<b>❌ Замечание:</b>

Чтобы не попасть в дамми-ловушку стоит удалить один из новых признаков geography и gender с помощью параметра drop_first=True
</div>

In [9]:
target = df_ohe['Exited']
features = df_ohe.drop(['Exited'], axis=1)

features_train, features_other, target_train, target_other = train_test_split(features, target, test_size=0.4, random_state=12345)
features_valid, features_test, target_valid, target_test = train_test_split(features_other, target_other, test_size=0.5, random_state=12345)

<div class="alert alert-block alert-success">
<b>✔️ Успех:</b> 

Данные поделены верно, пропорции выбраны адекватные 👍
</div>

## Исследование задачи

In [10]:
df_ohe['Exited'].value_counts()

0    7963
1    2037
Name: Exited, dtype: int64

<div style="margin: 10px 50px 10px 0px; padding: 20px; border: blue solid 1px ">
Классы не сбалансированны. Соотношение примерно 4:1.<br>
    Построим модели решающего древа и случайного леса на несбилансированных данных.
    </div>

<div class="alert alert-block alert-success">
<b>✔️ Успех:</b>

Виден явный дисбаланс классов в пользу отрицательного класса. Если мы будем всех людей прогнозировать, что они не уйдут из банка, то в 80% случаев мы будем правы.
</div>

In [11]:
# построим модиль для решающего древа

"""
Finds the best params for DecisionTreeClassifier withing the limit of number of leafs
"""
def tree_best_score(min_lim, max_lim, is_ballanced=False):
    best_score = 0
    best_depth = 0
    
    if is_ballanced:
        class_weight_ = 'balanced'
    else:
        class_weight_ = None
    
    for depth in range(min_lim, max_lim):
        model = DecisionTreeClassifier(max_depth=depth, class_weight=class_weight_, random_state=12345)
        model.fit(features_train, target_train)

        predicted_valid = model.predict(features_valid)
        f1_temp = f1_score(target_valid, predicted_valid)

        if f1_temp > best_score:
            best_score = f1_temp
            best_depth = depth
    return best_score, best_depth

tree_score, tree_depth = tree_best_score(2, 20)

print(f'Best score: {tree_score} best depth {tree_depth}')

Best score: 0.5714285714285714 best depth 9


In [12]:
tree_model = DecisionTreeClassifier(max_depth=9, random_state=12345)
tree_model.fit(features_train, target_train)

probabilities_valid = tree_model.predict_proba(features_valid)
# estimator.predict_proba(X, y)[:, 1]
tree_auc_roc = roc_auc_score(target_valid, probabilities_valid[:, 1])
print(tree_auc_roc)

0.7837430966797525


In [13]:
"""
Find the best params for RandomForest within the limits of estimators and depth
"""
def forest_best_score(min_depth_lim, max_depth_lim, min_est_lim, max_est_lim, est_step, is_ballanced=False):
    best_params = None
    best_result = 0
    
    if is_ballanced:
        class_weight_ = 'balanced'
    else:
        class_weight_ = None

    for depth in range(min_depth_lim, max_depth_lim):
        for est in range(min_est_lim, max_est_lim, est_step): 
            
            model = RandomForestClassifier(n_estimators=est, max_depth=depth, class_weight=class_weight_, random_state=12345)
            model.fit(features_train, target_train)

            predicted_valid = model.predict(features_valid)
            f1_temp = f1_score(target_valid, predicted_valid)

            if f1_temp > best_result:
                best_result = f1_temp
                best_params = est, depth
    
    return best_result, best_params

forest_score, forest_params = forest_best_score(2, 20, 20, 100, 10)
print(f'Best params n_est, depth: {forest_params} at score {forest_score}')       

Best params n_est, depth: (50, 17) at score 0.5878877400295421


In [14]:
forest_model = RandomForestClassifier(n_estimators=50, max_depth=17, class_weight=None, random_state=12345)
forest_model.fit(features_train, target_train)

probabilities_valid = forest_model.predict_proba(features_valid)
# estimator.predict_proba(X, y)[:, 1]
forest_auc_roc = roc_auc_score(target_valid, probabilities_valid[:, 1])
print(forest_auc_roc)

0.8321298519831357


<div style="margin: 10px 50px 10px 0px; padding: 20px; border: blue solid 1px ">
Мы попробовали подобрать параметры без учета дисбаланса для Решающего древа и для Случайного леса. <br>
Значение F1 меньше 0.59 <br>
Значения auc_roc 0.7837430966797525 и 0.8321298519831357 соответственно. <br>
    </div> 

<div class="alert alert-block alert-success">
<b>✔️ Успех:</b>

Модели обучены корректно 👍
</div>

## Борьба с дисбалансом

<div style="margin: 10px 50px 10px 0px; padding: 20px; border: blue solid 1px ">
Поборемся с дисбалансом при помощи балансировки класса
    </div>

In [15]:
# Сбалансированная модель Решающего Дерева

model_balanced = DecisionTreeClassifier(max_depth=9, class_weight='balanced', random_state=12345)
model_balanced.fit(features_train, target_train)

predicted_valid = model_balanced.predict(features_valid)
f1_result = f1_score(target_valid, predicted_valid)
print(f1_result)

0.5407554671968191


In [16]:
# Сбалансированная модель Случайного леса

model_balanced1 = RandomForestClassifier(n_estimators=50, max_depth=17, class_weight='balanced', random_state=12345)
model_balanced1.fit(features_train, target_train)

predicted_valid1 = model_balanced1.predict(features_valid)
f1_result1 = f1_score(target_valid, predicted_valid1)
print(f1_result1)

0.5671641791044776


In [17]:
# Проверка адекватности модели Случайного Леса

model_balanced1 = RandomForestClassifier(n_estimators=50, max_depth=17, class_weight='balanced', random_state=12345)
model_balanced1.fit(features_train, target_train)

predicted_valid1 = model_balanced1.predict(features_train)
f1_result1 = f1_score(target_train, predicted_valid1)
print(f1_result1)

0.997069903725408


In [18]:
forest_model = RandomForestClassifier(n_estimators=50, max_depth=17, class_weight='balanced', random_state=12345)
forest_model.fit(features_train, target_train)

probabilities_valid = forest_model.predict_proba(features_valid)
# estimator.predict_proba(X, y)[:, 1]
forest_auc_roc = roc_auc_score(target_valid, probabilities_valid[:, 1])
print(forest_auc_roc)

0.8415774956296616


<div style="margin: 10px 50px 10px 0px; padding: 20px; border: blue solid 1px ">
    Попробуем балансировать с помощью увеличения выборки
    </div>

In [19]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

<div class="alert alert-block alert-success">
<b>✔️ Успех:</b>

Увеличение выборки выполнено верно - классы сбалансированы 👍 
</div>

In [20]:
# Модель Дерва с увеличенной выборкой

model_ups = DecisionTreeClassifier(max_depth=9, random_state=12345)
model_ups.fit(features_upsampled, target_upsampled)

predicted_valid = model_ups.predict(features_valid)
f1_result = f1_score(target_valid, predicted_valid)
print(f1_result) 

0.5347912524850894


In [21]:
# Модель Леса с увеличенной выборкой

model_ups1 = RandomForestClassifier(n_estimators=50, max_depth=17, random_state=12345)
model_ups1.fit(features_upsampled, target_upsampled)

predicted_valid1 = model_ups1.predict(features_valid)
f1_result1 = f1_score(target_valid, predicted_valid1)
print(f1_result1)

0.6033376123234916


In [22]:
# Проверка адекватности Леса с увеличенной выборкой

model_ups1 = RandomForestClassifier(n_estimators=50, max_depth=17, random_state=12345)
model_ups1.fit(features_upsampled, target_upsampled)

predicted_valid1 = model_ups1.predict(features_train)
f1_result1 = f1_score(target_train, predicted_valid1)
print(f1_result1)

0.9979140592407176


In [23]:
forest_model = RandomForestClassifier(n_estimators=50, max_depth=17, random_state=12345)
forest_model.fit(features_upsampled, target_upsampled)

probabilities_valid = forest_model.predict_proba(features_valid)
# estimator.predict_proba(X, y)[:, 1]
forest_auc_roc = roc_auc_score(target_valid, probabilities_valid[:, 1])
print(forest_auc_roc)

0.8291560558677465


<div style="margin: 10px 50px 10px 0px; padding: 20px; border: blue solid 1px ">
После балансировки выглядит так что модель явно переобучена. <br>
    Далее я буду исследовать только модель случайного леса т.к. она дает более высокие показатели.<br>
    Подберем новые параметры для случайного леса с учетом балансировки
    </div>

<div class="alert alert-block alert-danger">
<b>❌ Замечание:</b>

Обычно под адекватностью модели имеют в виду сравнение его метрик с метрикой константной модели (например, которая предсказывает только один класс)

Если обучиться на увеличенной тренировочной выборке, затем давать предсказания на тренировочной выборке, то метрики, конечно, будут высокими) Тогда утверждать, что модель переобучилась не корректно. На валидационной выборке метрика адекватная - 0.6033376123234916
</div>

In [24]:
forest_score, forest_params = forest_best_score(2, 20, 20, 100, 10, is_ballanced=True)
print(f'Best params n_est, depth: {forest_params} at score {forest_score}')  

Best params n_est, depth: (70, 9) at score 0.6306306306306306


In [25]:
# Сбалансированная модель Леса с уточненными параметрами

model_balanced1 = RandomForestClassifier(n_estimators=70, max_depth=9, class_weight='balanced', random_state=12345)
model_balanced1.fit(features_train, target_train)

# n_estimators=100, max_depth=9

predicted_valid1 = model_balanced1.predict(features_valid)
f1_result1 = f1_score(target_valid, predicted_valid1)
print(f1_result1)

0.6306306306306306


In [26]:
# Проверка адекватности Леса с уточненными параметрами

model_ups1 = RandomForestClassifier(n_estimators=70, max_depth=9, random_state=12345)
model_ups1.fit(features_upsampled, target_upsampled)

predicted_valid1 = model_ups1.predict(features_train)
f1_result1 = f1_score(target_train, predicted_valid1)
print(f1_result1)

0.7478960848884011


In [27]:
forest_model = RandomForestClassifier(n_estimators=70, max_depth=9, class_weight='balanced', random_state=12345)
forest_model.fit(features_train, target_train)

probabilities_valid = forest_model.predict_proba(features_valid)
# estimator.predict_proba(X, y)[:, 1]
forest_auc_roc = roc_auc_score(target_valid, probabilities_valid[:, 1])
print(forest_auc_roc)

0.8512451684319408


In [28]:
# Модель Леса с уточненными параметрами с увеличенной выборкой

model_ups1 = RandomForestClassifier(n_estimators=70, max_depth=9, random_state=12345)
model_ups1.fit(features_upsampled, target_upsampled)

predicted_valid1 = model_ups1.predict(features_valid)
f1_result1 = f1_score(target_valid, predicted_valid1)
print(f1_result1)

0.6125


In [29]:
# Проверка адекватности Леса с уточненными параметрами с увеличенной выборкой

model_ups1 = RandomForestClassifier(n_estimators=70, max_depth=9, random_state=12345)
model_ups1.fit(features_upsampled, target_upsampled)

predicted_valid1 = model_ups1.predict(features_upsampled)
f1_result1 = f1_score(target_upsampled, predicted_valid1)
print(f1_result1)

0.8709917971662938


In [30]:
forest_model = RandomForestClassifier(n_estimators=70, max_depth=9, random_state=12345)
forest_model.fit(features_upsampled, target_upsampled)

probabilities_valid = forest_model.predict_proba(features_valid)
# estimator.predict_proba(X, y)[:, 1]
forest_auc_roc = roc_auc_score(target_valid, probabilities_valid[:, 1])
print(forest_auc_roc)

0.8495771810862635


<div style="margin: 10px 50px 10px 0px; padding: 20px; border: blue solid 1px ">
 При n_estimators=70, max_depth=9
    модель выглядит более адекватной: 0.6306306306306306. Наилучшие показатели для модели со сбалансированными классами. Приступим к ее тестированию.
    </div>

<div class="alert alert-block alert-warning">
<b>⚠️ Совет:</b>

В обучение можно добавить модель Логистической регрессии, чтобы сравнить линейную модель с моделями, основанными на деревьях решений
</div>

## Тестирование модели

In [31]:
model_test = RandomForestClassifier(n_estimators=70, max_depth=9, class_weight='balanced', random_state=12345)
model_test.fit(features_train, target_train)

predicted_test = model_test.predict(features_test)
f1_result = f1_score(target_test, predicted_test)
print(f1_result)

0.6077097505668935


In [32]:
forest_model = RandomForestClassifier(n_estimators=70, max_depth=9, class_weight='balanced', random_state=12345)
forest_model.fit(features_train, target_train)

probabilities_test = forest_model.predict_proba(features_test)
# estimator.predict_proba(X, y)[:, 1]
forest_auc_roc = roc_auc_score(target_test, probabilities_test[:, 1])
print(forest_auc_roc)

0.8553047576644768


<div class="alert alert-block alert-success">
<b>✔️ Успех:</b>

Отличный результат 👍 
</div>

<div class="alert alert-block alert-warning">
<b>⚠️ Совет:</b>

Здесь можно визуализировать ROC-кривую лучшей модели.
</div>

<div style="margin: 10px 50px 10px 0px; padding: 20px; border: blue solid 3px ">
<b> F1 метрика для случайного леса 0.6077097505668935 - что соответствует условиям. Так же значение auc_roc не значительно отличается от значения для валидационной выборки</b><br> <br> 
    <i>Мы получили модель, которая соответствует заявленным требованиям. <br> 
    Для этого мы обработали данные и заменили пропуски на медианные значения. <br> 
    Убрали колонки которые не влияют на модель. <br> 
    Закодировали категории чтобы значения в столбцах были числовыми. <br> 
    Разделили данные на обучающую валидационную и тестовую выборку. <br> 
    Испытали модель Дерева решений и Случайного леса.<br> 
    Создали модель и обучили ее на тренировочной выборки и на основе валидационной выборки подобрали лучшие параметры модели.<br> 
    Проверили дисбаланс и увидели что соотношение примерно 4:1<br> 
    Сбалансировали данные при момощи class_weight и увеличения выборки.<br> 
    Протестировали сбалансированные модели на выбранных параметрах. <br> 
    Выявили что модель переобучена. Подобрали новые параметры.<br> 
    Результат проверки модели на тестовой выборке показал что значение метрики f1 = 0.60 - что соответствует заявленным <br> требованиям. А метрика auc_roc 0.85 не значительно отличается от значения для валидационной выборки.</i><br> 
    <br> 
    </div>

<div class="alert alert-block alert-success">
<b>✔️ Успех:</b>

Вывод соответствует исследованию 👍
</div>

# <font color='orange'>Общее впечатление</font>
* Этот проект выполнен очень хорошо
* Видно, что приложено много усилий
* Молодец, что структурируешь ноутбук, приятно проверять такие работы
* У тебя чистый и лаконичный код
* Мне было интересно читать твои промежуточные выводы
* Твой уровень подачи материала находится на высоком уровне
* Исправь, пожалуйста, мои замечания. Затем отправляй на повторную проверку
* Жду новую версию проекта 👋

# <font color='orange'>Рекомендации 🔥</font>
* Анализ данных на python и pandas https://www.youtube.com/watch?v=dd3RcafdOQk&t=82s
* Курс "Введение в машинное обучение" https://stepik.org/course/4852
* Разумные способы кодирования категориальных данных для машинного обучения https://machinelearningmastery.ru/smarter-ways-to-encode-categorical-data-for-machine-learning-part-1-of-3-6dca2f71b159/
* Статья про "непонятную" метрику ROC-AUC https://dyakonov.org/2017/07/28/auc-roc-площадь-под-кривой-ошибок/
* Хочешь подтянуть математику для DS?) https://academy.stepik.org/math-for-data-science
* Результаты обучения моделей можно визуализировать 😎 https://www.datatechnotes.com/2019/08/elasticnet-regression-example-in-python.html
* В нашем деле нужно быть всегда в курсе всех новинок, новостей и тд, вот тут публикуют новости в области DS: https://www.infoq.com/data-analysis/news/
* Всем аналитикам данных рекомендую книгу Даниела Канемана "Думай медленно, решай быстро"